python/cuda_parallel/cuda/parallel/experimental/algorithms/reduce.py

```diff
@@ -97,43 +95,39 @@ def __call__(
         h_init: np.ndarray | GpuStruct,
         stream=None,
     ):
-        d_in_cccl = cccl.to_cccl_iter(d_in)
-        if d_in_cccl.type.value == cccl.IteratorKind.ITERATOR:
-            assert num_items is not None
+        if self.d_in_cccl.type.value == cccl.IteratorKind.POINTER:
+            self.d_in_cccl.state = protocols.get_data_pointer(d_in)
         else:
-            assert d_in_cccl.type.value == cccl.IteratorKind.POINTER
-            if num_items is None:
-                num_items = d_in.size
-            else:
-                assert num_items == d_in.size
-        _dtype_validation(
-            self._ctor_d_in_cccl_type_enum_name,
-            cccl.type_enum_as_name(d_in_cccl.value_type.type.value),
-        )
-        _dtype_validation(self._ctor_d_out_dtype, protocols.get_dtype(d_out))
-        _dtype_validation(self._ctor_init_dtype, h_init.dtype)
+            self.d_in_cccl.state = d_in.state
+
+        if self.d_out_cccl.type.value == cccl.IteratorKind.POINTER:
+            self.d_out_cccl.state = protocols.get_data_pointer(d_out)
+        else:
+            self.d_out_cccl.state = d_out.state
+
+        self.h_init_cccl.state = h_init.__array_interface__["data"][0]
```

Comment:
Q: I assume this is to get NumPy data pointer? If so, IIRC this is faster:
```suggestion
        self.h_init_cccl.state = h_init.ctypes.data
```

Reply:
Sorry, yes you're right... https://github.com/numpy/numpy/issues/23973#issuecomment-2177275191